velox/functions/prestosql/aggregates/ReduceAgg.cpp

```diff
@@ -80,40 +49,92 @@ using ReduceAggAccumulator = functions::aggregate::SingleValueAccumulator;
 //   s = f(s, 5)
 //   ...
 //
-//   The inputFunction lambda expression is evaluated 100 times, once per row.
+//   The inputFunction lambda expression was evaluated 100 times, once per row.
 //
-//   A more efficient approach would be:
+//   This implementation goes like this:
 //
 //   f - inputFunction, g - combineFunction
 //
-//   Convert all inputs into states:
+//   Convert all inputs into states using inputFunction and initialValue:
 //
 //   [s1, s2, s3, s4, s5,..s100] = f([s0, s0, s0, s0, s0,..], [1, 2, 3, 4, 5,..100])
 //
 //   Combine 100 states into 50:
 //
-//   [s1, s2,...s50] = g([s1, s3,...s99], [s2, s4,...s100])
+//   [s1, s2,...s50] = g([s1, s2,...s50], [s51, s52,...s100])
 //
 //   Combine these 50 states into 25 states:
 //
-//   [s1, s2,...s25] = g([s1, s3,...s49], [s2, s4,...s50])
+//   [s1, s2,...s25] = g([s1, s2,...s25], [s26, s27,...s50])
+//
+//   Combine these 25 states into 13 (12 + 1) states:
+//
+//   [s1, s2,...s12] = g([s1, s2,...s12], [s13, s14,...s24])
+//   s13 = s25
 //
 //   Continue in this manner until all states are combined. This requires only
 //   log2(100), ~7, expression evaluations.
```

Comment:
I guess what you mean here is that we only do 7 calls to the expression evaluation code, but that each call will process a larger number of rows/buckets/states, right? Took me a minute to understand

Reply:
Yes, that's what I mean. Do you have a rewrite in mind?